README.md
gst_plugins_tsunami
===================

Experimental (Tsunami) plugins


debian/changelog
mosami-gst-plugins-tsunami (0.1-1oneiric12) oneiric; urgency=low

  * Update objecttracker and add draweventbox plugins

 -- Paul HENRYS <Paul.Henrys@alcatel-lucent.com>  Fri, 21 Sep 2012 14:18:20 +0200

mosami-gst-plugins-tsunami (0.1-1oneiric11) oneiric; urgency=low

  * Rename predator plugin as objecttracker

 -- Paul HENRYS <Paul.Henrys@alcatel-lucent.com>  Wed, 19 Sep 2012 13:29:47 +0200

mosami-gst-plugins-tsunami (0.1-1oneiric10) oneiric; urgency=low

  * GCS, vibe and predator added and working

 -- Miguel Casas-Sanchez <miguel.casas-sanchez@alcatel-lucent.com>  Fri, 03 Aug 2012 10:35:10 +0200

mosami-gst-plugins-tsunami (0.1-1oneiric9) oneiric; urgency=low

  * Add predatortracker plugin

 -- Paul HENRYS <Paul.Henrys@alcatel-lucent.com>  Fri, 20 Jul 2012 16:06:53 +0200

mosami-gst-plugins-tsunami (0.1-1oneiric8) oneiric; urgency=low

  * Corrected again the dilate problem

 -- Miguel Casas-Sanchez <miguel.casas-sanchez@alcatel-lucent.com>  Thu, 19 Jul 2012 15:46:32 +0200

mosami-gst-plugins-tsunami (0.1-1oneiric7) oneiric; urgency=low

  * Added TSM and GCS plugins; Histeq changed to RGBA, "skin" added second method

 -- Miguel Casas-Sanchez <miguel.casas-sanchez@alcatel-lucent.com>  Thu, 19 Jul 2012 14:57:09 +0200

mosami-gst-plugins-tsunami (0.1-1oneiric6) oneiric; urgency=low

  * gstDilate -> gstDilate2 (conflic with a bad plugin)
    evolution of codebookfgbg (still experimental)

 -- Miguel Casas-Sanchez <miguel.casas-sanchez@alcatel-lucent.com>  Thu, 21 Jun 2012 19:09:30 +0200

mosami-gst-plugins-tsunami (0.1-1oneiric5) oneiric; urgency=low

  * #3256 passed some opencv and facetracker3 to rgb. Added ghosts

 -- Miguel Casas-Sanchez <miguel.casas-sanchez@alcatel-lucent.com>  Fri, 15 Jun 2012 15:37:53 +0200

mosami-gst-plugins-tsunami (0.1-1oneiric4) oneiric; urgency=low

  * repackage for oneiric

 -- Robin Nelissen <robin.nelissen@alcatel-lucent.com>  Fri, 13 Apr 2012 15:33:46 +0200

mosami-gst-plugins-tsunami (0.1-1dev16) lucid; urgency=low

  * modified histeq to do Y-norm with 1 of two methods

 -- Miguel Casas-Sanchez <miguel.casas-sanchez@alcatel-lucent.com>  Mon, 02 Apr 2012 17:17:57 +0200

mosami-gst-plugins-tsunami (0.1-1oneiric3) oneiric; urgency=low

  * add pixelate 

 -- Robin Nelissen <robin.nelissen@alcatel-lucent.com>  Wed, 28 Mar 2012 16:09:48 +0200

mosami-gst-plugins-tsunami (0.1-1dev15) lucid; urgency=low

  * add pixelate

 -- Robin Nelissen <robin.nelissen@alcatel-lucent.com>  Wed, 28 Mar 2012 14:55:11 +0200

mosami-gst-plugins-tsunami (0.1-1dev12) lucid; urgency=low

  * Removed zoom and vmix from tsunami

 -- Jan Van Winkel <jan.van_winkel@alcatel-lucent.com>  Tue, 27 Mar 2012 12:52:09 +0200

mosami-gst-plugins-tsunami (0.1-1oneiric2) oneiric; urgency=low

  * Removed zoom and vmix from tsunami 

 -- Jan Van Winkel <jan.van_winkel@alcatel-lucent.com>  Tue, 27 Mar 2012 12:42:02 +0200

mosami-gst-plugins-tsunami (0.1-1oneiric1) oneiric; urgency=low

  * Oneiric release 

 -- Jan Van Winkel <jan.van_winkel@alcatel-lucent.com>  Mon, 26 Mar 2012 18:45:31 +0200

mosami-gst-plugins-tsunami (0.1-1dev11) lucid; urgency=low

  * Corrected facetracke(missing symbols)

 -- Jan Van Winkel <jan.van_winkel@alcatel-lucent.com>  Mon, 26 Mar 2012 16:59:59 +0200

mosami-gst-plugins-tsunami (0.1-1dev10) lucid; urgency=low

  * Corrected makefile.am (opencv was missing)

 -- Jan Van Winkel <jan.van_winkel@alcatel-lucent.com>  Mon, 26 Mar 2012 14:30:11 +0200

mosami-gst-plugins-tsunami (0.1-1dev9) lucid; urgency=low

  * Moved facetracker from cyclops to tsunami 

 -- Jan Van Winkel <jan.van_winkel@alcatel-lucent.com>  Mon, 26 Mar 2012 11:26:46 +0200

mosami-gst-plugins-tsunami (0.1-1dev8) lucid; urgency=low

  * vmix->vmix0, zoom->zoom0

 -- Miguel Casas-Sanchez <miguel.casas-sanchez@alcatel-lucent.com>  Mon, 05 Mar 2012 14:54:51 +0100

mosami-gst-plugins-tsunami (0.1-1dev7) lucid; urgency=low

  * forcing opencv compilation

 -- Miguel Casas-Sanchez <miguel.casas-sanchez@alcatel-lucent.com>  Mon, 05 Mar 2012 11:16:02 +0100

mosami-gst-plugins-tsunami (0.1-1dev6) lucid; urgency=low

  * Sorted out minor path issues imco->mosami

 -- Miguel Casas-Sanchez <miguel.casas-sanchez@alcatel-lucent.com>  Mon, 05 Mar 2012 10:10:18 +0100

mosami-gst-plugins-tsunami (0.1-1dev5) lucid; urgency=low

  * trying again

 -- Miguel Casas-Sanchez <miguel.casas-sanchez@alcatel-lucent.com>  Thu, 01 Mar 2012 18:42:58 +0100

mosami-gst-plugins-tsunami (0.1-1dev4) lucid; urgency=low

  * trying agin

 -- Miguel Casas-Sanchez <miguel.casas-sanchez@alcatel-lucent.com>  Thu, 01 Mar 2012 18:19:58 +0100

mosami-gst-plugins-tsunami (0.1-1dev3) lucid; urgency=low

  * shlibs came clean, trying agin

 -- Miguel Casas-Sanchez <miguel.casas-sanchez@alcatel-lucent.com>  Thu, 01 Mar 2012 18:09:58 +0100

mosami-gst-plugins-tsunami (0.1-1dev2) lucid; urgency=low

  * Changed dependencies to mosami-*, and removed imcomsg dep (was in vmix)

 -- Miguel Casas-Sanchez <miguel.casas-sanchez@alcatel-lucent.com>  Thu, 01 Mar 2012 10:00:59 +0100

mosami-gst-plugins-tsunami (0.1-1dev1) lucid; urgency=low

  * First packaging of tsunami plugin

 -- Miguel Casas-Sanchez <miguel.casas-sanchez@alcatel-lucent.com>  Mon, 27 Feb 2012 14:00:20 +0100



src/vibe/vibe.c
#include "vibe.h"

#define VIBE_NRANDOM 7777

#include <string.h>

t_u_int8 expt[] = {0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
  		  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
		  0,   0,   0,   0,   0,   1,   1,   1,   1,   1,   1,   1,   1,   1,   1,   1,
		  1,   1,   1,   1,   1,   1,   1,   1,   1,   1,   1,   1,   1,   1,   1,   1,
		  2,   2,   2,   2,   2,   2,   2,   2,   2,   2,   2,   2,   2,   2,   2,   2,
		  3,   3,   3,   3,   3,   3,   3,   3,   3,   3,   3,   4,   4,   4,   4,   4,
		  4,   4,   4,   4,   5,   5,   5,   5,   5,   5,   5,   6,   6,   6,   6,   6,
		  6,   7,   7,   7,   7,   7,   7,   8,   8,   8,   8,   9,   9,   9,   9,   9,
		  10,  10,  10,   11,   11,   11,   11,   12,   12,   12,   13,   13,   13,   14,   14,   14,
		  15,  15,  16,   16,   16,   17,   17,   18,   18,   19,   19,   20,   20,   21,   21,   22,
		  23,  23,  24,   24,   25,   26,   26,   27,   28,   28,   29,   30,   31,   31,   32,   33,
		  34,  35,  36,   37,   38,   39,   40,   41,   42,   43,   44,   45,   46,   47,   49,   50,
		  51,  53,  54,   55,   57,   58,   60,   61,   63,   64,   66,   68,   70,   71,   73,   75,
		  77,  79,  81,    83,    85,    88,    90,    92,    94,    97,    99,   102,   105,   107,   110,   113,
		  116, 119, 122,   125,   128,   132,   135,   138,   142,   146,   149,   153,   157,   161,   165,   170,
		  174, 178, 183,   188,  193,   197,   203,   208,   213,   219,   224,   230,   236,  242,   248,   255};


t_vibe *vibe_create(t_u_int32 width, t_u_int32 height, t_u_int8 nsamples)
{
  t_vibe *vb = (t_vibe*)malloc(sizeof(t_vibe));
  t_u_int32 i;

  vb->width = width;
  vb->height = height;
  vb->nsamples = nsamples;

  vb->cardinality = nsamples/3;
  vb->sigma_y = 24;
  vb->sigma_u = 12;
  vb->sigma_v = 12;

  vb->lspeed_fg = 80;//4096;
  vb->lspeed_bg = 10; //128;

  vb->cdec_fg = 1;
  vb->cinc_bg = 100;

  vb->lastscore = 0.5;

  vb->random = (t_u_int32*)malloc(VIBE_NRANDOM*sizeof(t_u_int32));
  vb->irandom = 0;
  for (i=0; i<VIBE_NRANDOM; i++) vb->random[i] = rand()%255;

  vb->model = (t_u_int8*)malloc(3*vb->width*vb->height*vb->nsamples*sizeof(t_u_int8));
  memset(vb->model,0,3*vb->width*vb->height*vb->nsamples*sizeof(t_u_int8));
  vb->conf = (t_u_int8*)malloc(vb->width*vb->height*sizeof(t_u_int8));
  memset(vb->conf,0,vb->width*vb->height*sizeof(t_u_int8));

  return vb;
}
void vibe_destroy(t_vibe *vb) 
{
  free(vb->model);
  free(vb->conf);
  free(vb->random);
  free(vb);
}

t_u_int32 vibe_getrandom(t_vibe *vb) {
  t_u_int32 r = vb->random[vb->irandom];
  vb->irandom++; 
  if (vb->irandom==VIBE_NRANDOM) 
    vb->irandom=0;
  return r;
}

// jvw mcs 14/aug/12
// per pixel we need to score the image agains the model, meaning compare the
// distribution parameters yuv between both and if the relative values are
// larger than a certain threshold we admit it has changed (fg) otherwise is
// not changed, so is background. The vibe->model is thus having values 
// {0, alpha, 255}, where alpha is sigma* alpha where alpha \in [-0.5,0.5]
// and sigma is a constant from the beginning
void vibe_initmodel(t_vibe *vb, IplImage *image)
{
  t_u_int32 i, j, k;

  t_u_int8 *mm = vb->model;
  for(i=0; i<image->height; i++)
  {
    t_u_int8 *ii = (t_u_int8*)(image->imageData + image->widthStep*i);		
    for (j=0; j<image->width; j++, ii+=4)
    {
      for (k=0; k<vb->nsamples; k++, mm+=3)
      {
        int m;
        m = ii[0] + ((int)vb->sigma_y * ((int)(vibe_getrandom(vb) % 256) - 127))/256;
        mm[0]=m; 
        if (m<0) 
          mm[0]=0; 
        else if (m>255) 
          mm[0]=255;
        m = ii[1] + ((int)vb->sigma_u * ((int)(vibe_getrandom(vb) % 256) - 127))/256;
        mm[1]=m; 
        if (m<0) 
          mm[1]=0; 
        else if (m>255) 
          mm[1]=255;
        m = ii[2] + ((int)vb->sigma_v * ((int)(vibe_getrandom(vb) % 256) - 127))/256;
        mm[2]=m; 
        if (m<0) 
          mm[2]=0; 
        else if (m>255) 
          mm[2]=255;
      }
    }
  }

  memset(vb->conf,255,vb->width*vb->height*sizeof(t_u_int8));
}

void vibe_segment(t_vibe *vb, IplImage *image)
{
  t_u_int32 i, j, k;
  t_u_int8 card;
  t_double totscore;

  t_u_int32 suv = vb->sigma_u * vb->sigma_v; suv *= suv;
  t_u_int32 svy = vb->sigma_v * vb->sigma_y; svy *= svy;
  t_u_int32 syu = vb->sigma_y * vb->sigma_u; syu *= syu;
  t_u_int32 syuv = vb->sigma_y * vb->sigma_u * vb->sigma_v; syuv *= syuv;

  t_u_int8 *mm = vb->model;
  t_u_int8 *cc = vb->conf;
  for(i=0, totscore = 0.0; i<image->height; i++)
  {
    t_u_int8 *ii = (t_u_int8*)image->imageData + image->widthStep*i;		
    for (j=0; j<image->width; j++, ii+=4, mm+=3*vb->nsamples, cc++)
    {
      //ii[3] = 255;
      for (k=0, card=0; k<vb->nsamples; k++)
      {
        t_s_int32 dy = mm[3*k]  -ii[0];
        t_s_int32 du = mm[3*k+1]-ii[1];
        t_s_int32 dv = mm[3*k+2]-ii[2];
        if (dy*dy*suv + du*du*svy + dv*dv*syu < syuv) card++;
        //if (card >= vb->cardinality) {
        //	ii[3] = 0;
        //	break;
        //}
      }
			
      //t_s_int32 score = (vb->nsamples-card)*40;
      t_s_int32 score = (cc[0] * (2*vb->cardinality - card))/(2*vb->cardinality) + 127 - cc[0]/2;
      ii[3] = (score>0)? score : 0;

      totscore += ii[3];
    }
  }

  totscore /= (image->width*image->height*255.0);
	
  if (totscore-vb->lastscore > 0.1) {
    printf("[Vibe] camera movement (%f/%f), resetting confidence \n", totscore, vb->lastscore);
    memset(vb->conf,0,vb->width*vb->height*sizeof(t_u_int8));
    for(i=0; i<image->height; i++)
    {
      t_u_int8 *ii = (t_u_int8*)image->imageData + image->widthStep*i;		
      for (j=0; j<image->width; j++, ii+=4)
      {
        ii[3] = 127;
        totscore += ii[3];
      }
    }
    totscore = 0.5;
  }
	

  vb->lastscore = totscore;
}

// jvw mcs 14/aug/12
// Update one model which is randomly chosen, using t, a decay factor
// calculated using lspeed_fg and l_speed_bg
void vibe_update(t_vibe *vb, IplImage *image)
{
  t_u_int32 i, j;

  t_u_int8 *mm = vb->model;
  t_u_int8 *cc = vb->conf;
  for(i=0; i<image->height; i++)
  {
    t_u_int8 *ii = (t_u_int8*)image->imageData + image->widthStep*i;		
    for (j=0; j<image->width; j++, ii+=4, mm+=3*vb->nsamples, cc++) {
      //t_s_int32 t = (ii[3]*vb->lspeed_fg)/255 + ((255-ii[3])*vb->lspeed_bg)/255;
      t_s_int32 t = (expt[ (ii[3]*cc[0])/255 ] * vb->lspeed_fg)/255.0 + vb->lspeed_bg;
      //t = (t*cc[0])/255;
      if ((t<2)||(vibe_getrandom(vb) % t ==0)) { 
        t_u_int32 r = vibe_getrandom(vb) % vb->nsamples;
        mm[3*r] = ii[0];
        mm[3*r+1] = ii[1];
        mm[3*r+2] = ii[2];
        cc[0] = (cc[0]<256-vb->cinc_bg)? cc[0]+vb->cinc_bg : 255;
      } else {
        cc[0] = (cc[0]>vb->cdec_fg-1)? cc[0]-vb->cdec_fg : 0;
      }
    }
  }
}

void vibe_display(t_vibe *vb, IplImage *image)
{
  t_u_int32 i, j;

  t_u_int8 *cc = vb->conf;
  for(i=0; i<image->height; i++)
  {
    t_u_int8 *ii = (t_u_int8*)image->imageData + image->widthStep*i;		
    for (j=0; j<image->width; j++, ii+=4, cc++) {
      ii[0] = ii[3];
      ii[1] = 127; //ii[3];
      ii[2] = 127; //ii[3];
    }
  }
}

void vibe_display_model(t_vibe *vb, IplImage *image)
{
  t_u_int32 i, j, k;

  t_u_int16 sum;

  t_u_int8 *mm = vb->model;
  t_u_int8 *cc = vb->conf;
  for(i=0; i<image->height; i++)
  {
    t_u_int8 *ii = (t_u_int8*)image->imageData + image->widthStep*i;		
    for (j=0; j<image->width; j++, ii+=4, cc++, mm+=3*vb->nsamples) {
      for (k=0,sum=0; k<vb->nsamples;k++) 
        sum+=mm[3*k]; 
      ii[0] = sum/vb->nsamples;
      for (k=0,sum=0; k<vb->nsamples;k++) 
        sum+=mm[3*k+1]; 
      ii[1] = sum/vb->nsamples;
      for (k=0,sum=0; k<vb->nsamples;k++) 
        sum+=mm[3*k+2]; 
      ii[2] = sum/vb->nsamples;
    }
  }
}
